kamu-adapter-http/src/ws_smart_transfer_protocol_axum_server.rs

```diff
@@ -318,11 +350,251 @@ pub async fn dataset_pull_ws_handler(
+                push_metadata_request.new_blocks.media_type,
+                push_metadata_request.new_blocks.encoding,
+                push_metadata_request.new_blocks.payload.len()
+            );
+
+            assert!(
+                push_request.size_estimation.num_blocks
+                    == push_metadata_request.new_blocks.objects_count
+            );
+
+            let new_blocks =
+                dataset_import_pushed_metadata(dataset, push_metadata_request.new_blocks).await;
+
+            let response_result_metadata = write_payload::<DatasetPushMetadataAccepted>(
+                socket,
+                DatasetPushMetadataAccepted {},
+            )
+            .await;
+
+            if let Err(e) = response_result_metadata {
+                Err(SmartProtocolPushServerError::PushMetadataResponseWriteFailed(e))
+            } else {
+                Ok(new_blocks)
+            }
+        }
+        Err(ReadMessageError::Closed) => Ok(vec![]),
+        Err(e) => Err(SmartProtocolPushServerError::PushMetadataRequestReadFailed(
+            e,
+        )),
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////
+
+async fn try_handle_push_objects_request(
+    socket: &mut axum::extract::ws::WebSocket,
+    dataset: &dyn Dataset,
+    dataset_url: &Url,
+) -> Result<bool, SmartProtocolPushServerError> {
+    let maybe_push_objects_transfer_request =
+        read_payload::<DatasetPushObjectsTransferRequest>(socket).await;
+
+    match maybe_push_objects_transfer_request {
+        Ok(request) => {
+            tracing::debug!(
+                "Push client sent a push objects request for {} objects",
+                request.object_files.len(),
+            );
+
+            let mut object_transfer_strategies: Vec<PushObjectTransferStrategy> = Vec::new();
+            for r in request.object_files {
+                let transfer_strategy =
+                    prepare_push_object_transfer_strategy(dataset, &r, &dataset_url).await;
+
+                match transfer_strategy {
+                    Ok(strategy) => object_transfer_strategies.push(strategy),
+                    Err(e) => return Err(SmartProtocolPushServerError::Internal(e)),
+                };
+            }
+
+            tracing::debug!("Object transfer strategies defined");
+
+            let response_result_objects = write_payload::<DatasetPushObjectsTransferResponse>(
+                socket,
+                DatasetPushObjectsTransferResponse {
+                    object_transfer_strategies,
+                },
+            )
+            .await;
+
+            if let Err(e) = response_result_objects {
+                Err(SmartProtocolPushServerError::PushObjectResponseWriteFailed(
+                    e,
+                ))
+            } else {
+                Ok(request.is_truncated)
+            }
+        }
+        Err(e) => Err(SmartProtocolPushServerError::PushObjectRequestReadFailed(e)),
```

Comment:
Lots of nested `match` blocks are kinda hard to read, so I would recommend using `map_err(...)?` when you only match for error handling and do early returns:

```rust
let maybe_push_objects_transfer_request =
  read_payload::<DatasetPushObjectsTransferRequest>(socket)
  .await
  .map_err(...)?;
```

Reply:
Love this idiom. I've simplified a lot of code with `map_err`